Enable test timeout dump collection

marcpopMSFT opened the pull request:
I do not know the history on why we launch our tests with dotnet-exec. We may hae to change the arguments to match the above commands that we use on netfx (and simplify to just use one).

I'm starting with this just to see if it'll work.

Also open question is where will the dump end up.

marcpopMSFT committed:
Enable test timeout dump collection

marcpopMSFT committed:
Simplify the logic for the test command since we're going to use dotnet test for both full framework and not

marcpopMSFT committed:
add missing double quote

marcpopMSFT committed:
Remove the netfx conditional for the scheduler filter so it works with the dotnet test format

marcpopMSFT committed:
Fix the build breaks

marcpopMSFT committed:
try fixing the path parameter

marcpopMSFT commented:
This doesn't appear to work on non-windows:
`The test source file "/private/tmp/helix/working/AB8E0979/w/B5F809E7/e/trx" provided was not found.`

And any test that requires msbuildforwardingapp doesn't appear to work correctly within the dotnet test execution.

Applying a fix for the path variable as it was adding quotes that weren't meant to be there.

marcpopMSFT commented:
@dsplaisted some of the other failures doing this are because we end up getting the host from the environment but that's now testhost.exe: https://github.com/dotnet/sdk/blob/main/src/Cli/Microsoft.DotNet.Cli.Utils/ForwardingAppImplementation.cs#L99

Is there a better way to find dotnet.exe that we already have in our test infrastructure?

marcpopMSFT committed:
Add output to the failure message as it's not captured in helix output

am11 commented:
illink (and few other tools in runtime repo) finds it via `DOTNET_HOST_PATH` https://github.com/dotnet/runtime/blob/eee9fd32dbf29d4df263c940b394a8af47bd5cf5/src/tools/illink/src/ILLink.Tasks/LinkTask.cs#L235

marcpopMSFT commented:
will keep rerunning tests here to see if we can reproduce the arm64 macos test timeout while also looking for a solutino to the path problem. Using an env variable as suggested might work but I was looking for a standard option.

marcpopMSFT committed:
Let's try falling back to host path and if that's not set, let's dump the environment block so we can see what else we can potentially use.

marcpopMSFT committed:
Fall back to DOTNET_ROOT if HOST_PATH isn't set.

marcpopMSFT committed:
Add dotnet to the path itself.

marcpopMSFT committed:
Fall back to host and root paths for finding the muxer

marcpopMSFT commented:
@dsplaisted let's chat on Monday about this. I got it working but in two places I had to change product code to fall back to env variables to find the dotnet host when not running. Not sure if we feel comfortable with that and would rather target 9.0.2xx since it's technically customer impacting now though I'm not sure how those scenarios would have worked before if they were returning the testhost.

am11 commented on src/Cli/Microsoft.DotNet.Cli.Utils/ForwardingAppImplementation.cs:
```diff
@@ -96,7 +97,19 @@ public ForwardingAppImplementation WithEnvironmentVariable(string name, string v
         private string GetHostExeName()
         {
             // Should instead make this a full path to dotnet
-            return Environment.ProcessPath;
+            string processPath = Environment.ProcessPath;
+            string hostName = "dotnet" + Constants.ExeSuffix;
+
+            if (!processPath.EndsWith(hostName, StringComparison.OrdinalIgnoreCase))
+            {
+                var dotnetHostPath = Environment.GetEnvironmentVariable("DOTNET_HOST_PATH") ?? Environment.GetEnvironmentVariable("DOTNET_ROOT");
+                if (!string.IsNullOrEmpty(dotnetHostPath))
+                {
+                    processPath = dotnetHostPath+ Path.DirectorySeparatorChar + hostName;
+                }
+            }
+
+            return processPath;
         }
```
```suggestion
        private string GetHostExeName() => new Muxer().MuxerPath;
```
(`MuxerPath` property can be static to avoid new'ing the instance)

am11 commented on src/Cli/Microsoft.DotNet.Cli.Utils/Muxer.cs:
```diff
@@ -35,7 +35,18 @@ public string MuxerPath
         public Muxer()
         {
 #if NET6_0_OR_GREATER
-            _muxerPath = Environment.ProcessPath;
+            string processPath = Environment.ProcessPath;
+            string hostName = "dotnet" + Constants.ExeSuffix;
+
+            if (!processPath.EndsWith(hostName, StringComparison.OrdinalIgnoreCase))
+            {
+                var dotnetHostPath = Environment.GetEnvironmentVariable("DOTNET_HOST_PATH") ?? Environment.GetEnvironmentVariable("DOTNET_ROOT");
+                if (!string.IsNullOrEmpty(dotnetHostPath))
+                {
+                    processPath = dotnetHostPath + Path.DirectorySeparatorChar + hostName;
+                }
+            }
+            _muxerPath = processPath;
```
```suggestion
#if NET
            // SDK sets DOTNET_HOST_PATH as absolute path to current dotnet executable
            string processPath = Environment.GetEnvironmentVariable("DOTNET_HOST_PATH");
            if (string.IsNullOrEmpty(processPath))
            {
                // fallback to DOTNET_ROOT which is guaranteed to hold some dotnet executable
                var root = Environment.GetEnvironmentVariable("DOTNET_ROOT");
                if (!string.IsNullOrEmpty(root))
                {
                    processPath = Path.Combine(root, $"dotnet{Constants.ExeSuffix}");
                }
                else
                {
                    // this is a wildcard, the executing process may or may not be dotnet
                    string processPath = Environment.ProcessPath;
                }
            }
```

am11 reviewed (commented):
@marcpopMSFT I think reliability-wise, this is the correct order. Note: `DOTNET_HOST_PATH` holds the absolute path to executable (unlike `DOTNET_ROOT` which points to directory)

am11 commented on src/Cli/Microsoft.DotNet.Cli.Utils/ForwardingAppImplementation.cs:
```diff
@@ -4,6 +4,7 @@
 #if NET
 
 using System.Diagnostics;
+using System.Collections;
```
```suggestion
```

am11 commented on src/Cli/Microsoft.DotNet.Cli.Utils/Muxer.cs:
```diff
@@ -35,7 +35,18 @@ public string MuxerPath
         public Muxer()
         {
 #if NET6_0_OR_GREATER
-            _muxerPath = Environment.ProcessPath;
+            string processPath = Environment.ProcessPath;
+            string hostName = "dotnet" + Constants.ExeSuffix;
+
+            if (!processPath.EndsWith(hostName, StringComparison.OrdinalIgnoreCase))
+            {
+                var dotnetHostPath = Environment.GetEnvironmentVariable("DOTNET_HOST_PATH") ?? Environment.GetEnvironmentVariable("DOTNET_ROOT");
+                if (!string.IsNullOrEmpty(dotnetHostPath))
+                {
+                    processPath = dotnetHostPath + Path.DirectorySeparatorChar + hostName;
+                }
+            }
+            _muxerPath = processPath;
 #else
             _muxerPath = Process.GetCurrentProcess().MainModule.FileName;
 #endif
```
```suggestion
        public Muxer()
        {
            // SDK sets DOTNET_HOST_PATH as absolute path to current dotnet executable
            string processPath = Environment.GetEnvironmentVariable("DOTNET_HOST_PATH");
            if (string.IsNullOrEmpty(processPath))
            {
                // fallback to DOTNET_ROOT which typically holds some dotnet executable
                var root = Environment.GetEnvironmentVariable("DOTNET_ROOT");
                if (!string.IsNullOrEmpty(root))
                {
                    processPath = Path.Combine(root, $"dotnet{Constants.ExeSuffix}");
                }
                else
                {
                    // this is a wildcard, the executing process may or may not be dotnet
#if NET
                    processPath = Environment.ProcessPath;
#else
                    processPath = Process.GetCurrentProcess().MainModule.FileName;
#endif
                }
            }

            _muxerPath = processPath;
```

dsplaisted replied:
For this comment:

```
// this is a wildcard, the executing process may or may not be dotnet
```

I think that when running the product, the process will always be dotnet.  It is only when loading dotnet.dll to run tests that it could be a different process.  So I would update the comment to reflect that.

am11 replied:
@marcpopMSFT, please incorporate "DOTNET_HOST_PATH holds the absolute path to executable (unlike DOTNET_ROOT which points to directory)". Appending dotnet.exe to `DOTNET_HOST_PATH` will result in incorrect path. e.g. with daily build and -v:diag, I get `DOTNET_HOST_PATH = /Users/am11/.dotnet10/dotnet`

marcpopMSFT replied:
hmm, I'm surprised the PR was passing checks if that was the case but from a few binlogs I found, that appears to be accurate. I'll make the change and let's see if the checks pass this time.

marcpopMSFT committed:
Apply PR feedback for most cases.

For the Muxer logic itself, we probably want to stick to only using HOST_PATH when we detect the current process is not the muxer so as to reduce risk to the product.

am11 commented on src/Cli/Microsoft.DotNet.Cli.Utils/Muxer.cs:
```diff
@@ -34,11 +34,26 @@ public string MuxerPath
+            // The current process should be dotnet in most normal scenarios except when dotnet.dll is loaded in a custom host like the testhost
+            if (!processPath.EndsWith(hostName, StringComparison.OrdinalIgnoreCase))
+            {
+                var dotnetHostPath = Environment.GetEnvironmentVariable("DOTNET_HOST_PATH") ?? Environment.GetEnvironmentVariable("DOTNET_ROOT");
+                if (!string.IsNullOrEmpty(dotnetHostPath))
+                {
+                    processPath = dotnetHostPath + Path.DirectorySeparatorChar + hostName;
+                }
```
```suggestion
	            // SDK sets DOTNET_HOST_PATH as absolute path to current dotnet executable
	            processPath = Environment.GetEnvironmentVariable("DOTNET_HOST_PATH");
	            if (processPath is null)
	            {
	                // fallback to DOTNET_ROOT which typically holds some dotnet executable
	                var root = Environment.GetEnvironmentVariable("DOTNET_ROOT");
	                if (root is not null)
	                {
	                    processPath = Path.Combine(root, $"dotnet{Constants.ExeSuffix}");
	                }
	            }
```

marcpopMSFT committed:
Merge branch 'release/9.0.1xx' into marcpopMSFT-collectdump

marcpopMSFT committed:
Only a

Co-authored-by: Adeel Mujahid <3840695+am11@users.noreply.github.com>

marcpopMSFT commented:
BTW, I checked the test results and locally and it looks like using dotnet test with xunit in helix will report the test work item as an additional failure. This was old behavior we used to have for the dotnet exec method of launching tests which we didn't like as it double counted failures and made the AzDO UI less user friendly.

I'm going to ping the test folks to see if there is an option for this.

dsplaisted commented on src/Cli/Microsoft.DotNet.Cli.Utils/Muxer.cs:
```diff
@@ -34,11 +34,32 @@ public string MuxerPath
+            string hostName = "dotnet" + Constants.ExeSuffix;
+
+            // The current process should be dotnet in most normal scenarios except when dotnet.dll is loaded in a custom host like the testhost
+            if (!processPath.EndsWith(hostName, StringComparison.OrdinalIgnoreCase))
```
This should fix the case where the process name is `notdotnet.exe`:

```suggestion
            if (!Path.GetFileNameWithoutExtension(processPath).Equals("dotnet", StringComparison.OrdinalIgnoreCase))
```

marcpopMSFT committed:
Use GetFileNamewithoutExtension

Co-authored-by: Daniel Plaisted <daplaist@microsoft.com>

am11 commented on src/Cli/Microsoft.DotNet.Cli.Utils/Muxer.cs:
```diff
@@ -34,11 +34,32 @@ public string MuxerPath
-            _muxerPath = Process.GetCurrentProcess().MainModule.FileName;
+            string processPath = Process.GetCurrentProcess().MainModule.FileName;
 #endif
+            string hostName = "dotnet" + Constants.ExeSuffix;
```
```suggestion
```
(unused)

marcpopMSFT committed:
remove unused hostname

Co-authored-by: Adeel Mujahid <3840695+am11@users.noreply.github.com>

marcpopMSFT committed:
Merge branch 'release/9.0.1xx' into marcpopMSFT-collectdump

marcpopMSFT commented:
unrelated failure in `Microsoft.DotNet.Watcher.Tools.FileWatcherTests.DeleteSubfolder`. Rerrunning all to see if that resolves it or else I'll ping the watch team next. Don't see the failure in other PRs that I've poked at.
